### PR TITLE
ARROW-3853: [C++] Implement string to timestamp cast

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Comment by pitrou** [comment]
Micro-benchmark (AMD Ryzen 7, gcc 7.3.0):
```
-----------------------------------------------------------------------------
Benchmark                                      Time           CPU Iterations
-----------------------------------------------------------------------------
BM_IntegerParsing<Int8Type>                 2977 ns       2976 ns     234204    320.41M items/s
BM_IntegerParsing<Int16Type>                3227 ns       3226 ns     216686   295.581M items/s
BM_IntegerParsing<Int32Type>                3907 ns       3906 ns     177879   244.144M items/s
BM_IntegerParsing<Int64Type>                5989 ns       5988 ns     112958   159.264M items/s
BM_IntegerParsing<UInt8Type>                1810 ns       1809 ns     379162   527.129M items/s
BM_IntegerParsing<UInt16Type>               2007 ns       2006 ns     350245   475.319M items/s
BM_IntegerParsing<UInt32Type>               2400 ns       2399 ns     292529   397.491M items/s
BM_IntegerParsing<UInt64Type>               3672 ns       3671 ns     188921   259.795M items/s
BM_FloatParsing<FloatType>                 66782 ns      66771 ns      10158   14.2829M items/s
BM_FloatParsing<DoubleType>                41465 ns      41458 ns      16674   23.0033M items/s
BM_TimestampParsing<TimeUnit::SECOND>      13649 ns      13646 ns      50778   69.8856M items/s
BM_TimestampParsing<TimeUnit::MILLI>       13952 ns      13950 ns      49803    68.364M items/s
BM_TimestampParsing<TimeUnit::MICRO>       13736 ns      13733 ns      49687   69.4425M items/s
BM_TimestampParsing<TimeUnit::NANO>        14019 ns      14017 ns      50912   68.0377M items/s

```

File: cpp/src/arrow/compute/kernels/cast.cc [hunk 1]
```diff
@@ -980,6 +980,36 @@ struct CastFunctor<O, StringType,
...
 };
 
+// ----------------------------------------------------------------------
+// String to Timestamp
+
+template <>
+struct CastFunctor<TimestampType, StringType> {
+  void operator()(FunctionContext* ctx, const CastOptions& options,
+                  const ArrayData& input, ArrayData* output) {
+    //     const auto& out_type = checked_cast<const TimestampType&>(*output->type);
```
Comment (fsaintjacques) [inline cpp/src/arrow/compute/kernels/cast.cc]: Leftover.

File: cpp/src/arrow/compute/kernels/cast.cc [hunk 2]
```diff
@@ -980,6 +980,36 @@ struct CastFunctor<O, StringType,
...
+    using out_type = TimestampType::c_type;
+
+    StringArray input_array(input.Copy());
```
Comment (fsaintjacques) [inline cpp/src/arrow/compute/kernels/cast.cc]: Why does it need to copy?

File: cpp/src/arrow/compute/kernels/cast.cc [hunk 3]
```diff
@@ -980,6 +980,36 @@ struct CastFunctor<O, StringType,
...
+    auto out_data = output->GetMutableValues<out_type>(1);
+    internal::StringConverter<TimestampType> converter(output->type);
+
+    for (int64_t i = 0; i < input.length; ++i, ++out_data) {
+      if (input_array.IsNull(i)) {
+        continue;
```
Comment (fsaintjacques) [inline cpp/src/arrow/compute/kernels/cast.cc]: Should you push a null to out_data?

File: cpp/src/arrow/util/parsing.h [hunk 4]
```diff
@@ -419,8 +419,9 @@ class StringConverter<TimestampType> {
         *out = std::chrono::duration_cast<std::chrono::nanoseconds>(duration).count();
         return true;
     }
-    // Unreachable
+    // Unreachable, but suppress compiler warning
     assert(0);
+    *out = 0;
```
Comment (fsaintjacques) [inline cpp/src/arrow/util/parsing.h]: I'd set this in the `default` case, this is a bit friendlier to callers if they don't check the result.

File: cpp/src/arrow/compute/compute-test.cc [hunk 5]
```diff
@@ -1056,6 +1056,37 @@ TEST_F(TestCast, StringToNumberErrors) {
   CheckFails<StringType, std::string>(utf8(), {"z"}, is_valid, float32(), options);
 }
 
+TEST_F(TestCast, StringToTimestamp) {
+  CastOptions options;
+
+  vector<bool> is_valid = {true, false, true, true};
```
Comment (fsaintjacques) [inline cpp/src/arrow/compute/compute-test.cc]: Does not match `strings` size. Worth adding a DCHECK in ArrayFromVector.

File: cpp/src/arrow/compute/kernels/cast.cc [hunk 6]
```diff
@@ -980,6 +980,36 @@ struct CastFunctor<O, StringType,
...
+      }
+
+      auto str = input_array.GetView(i);
+      if (!converter(str.data(), str.length(), out_data)) {
```
Comment (fsaintjacques) [inline cpp/src/arrow/compute/kernels/cast.cc]: I think you need to set the is valid bit. 

I would add a way to skip errors (safe/unsafe) as opposed to bork. This become very annoying in huge 
(always broken in some way) datasets.

File: cpp/src/arrow/compute/kernels/cast.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/compute/kernels/cast.cc]: Right, thank you.

File: cpp/src/arrow/compute/kernels/cast.cc (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/compute/kernels/cast.cc]: This is a shallow copy (it doesn't copy the underlying data, just duplicate the containers). But good question. I took it from the other casting implementations. Perhaps constructing an Array may mutate the underlying ArrayData and we don't want that? @wesm

File: cpp/src/arrow/compute/kernels/cast.cc (hunk 3 above)
Comment (pitrou) [reply cpp/src/arrow/compute/kernels/cast.cc]: I think nulls are simply copied as a separate pass.

File: cpp/src/arrow/compute/kernels/cast.cc (hunk 2 above)
Comment (wesm) [reply cpp/src/arrow/compute/kernels/cast.cc]: NB: the copy is shallow / the memory buffers are not copied (I am not sure why it is needed though)

File: cpp/src/arrow/compute/kernels/cast.cc (hunk 6 above)
Comment (pitrou) [reply cpp/src/arrow/compute/kernels/cast.cc]: I wouldn't be surprised if there's already a JIRA open for a "nulls on errors" option :-)

File: cpp/src/arrow/util/parsing.h (hunk 4 above)
Comment (pitrou) [reply cpp/src/arrow/util/parsing.h]: Hmm... what do you mean? It should do the same in the end.

File: cpp/src/arrow/compute/kernels/cast.cc (hunk 6 above)
Comment (fsaintjacques) [reply cpp/src/arrow/compute/kernels/cast.cc]: Some casts do it.

I'd also propose to simplify the whole CastOptions, to

- Toggling stop on errors (safe vs unsafe), call it "allow_truncation".
- Truncation mode: set bit to null _or_ default to a defined value, e.g. `0` or `INT64_MIN`, "NA", etc.

 @wesm what do you think?

File: cpp/src/arrow/compute/compute-test.cc (hunk 5 above)
Comment (pitrou) [reply cpp/src/arrow/compute/compute-test.cc]: Hmm, good catch.

File: cpp/src/arrow/compute/kernels/cast.cc (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/compute/kernels/cast.cc]: Yes.

Comment (pitrou) [reply cpp/src/arrow/compute/kernels/cast.cc]: As for the copy, actually it's simple: `Array` expects a `shared_ptr<ArrayData>`, and we have a bare reference here, from which we cannot safely devise a `shared_ptr` unless we copy it...

File: cpp/src/arrow/util/parsing.h (hunk 4 above)
Comment (fsaintjacques) [reply cpp/src/arrow/util/parsing.h]: Maybe it's just an old habit of using `-Wswitch-enum`.

**Comment by pitrou** [comment]
The Travis-CI failure is due to persistent Anaconda outage.

File: cpp/src/arrow/compute/kernels/cast.cc (hunk 6 above)
Comment (wesm) [reply cpp/src/arrow/compute/kernels/cast.cc]: https://issues.apache.org/jira/browse/ARROW-1489

**Comment by pitrou** [comment]
Rebased.

**Comment by fsaintjacques** [comment]
LGTM

**Review by wesm — APPROVED** [review]
+1

**Pull request closed without merging**